channels_redis/alt.py

```diff
@@ -0,0 +1,311 @@
+import asyncio
+import sys
+import traceback
+import uuid
+
+import aioredis
+import msgpack
+
+
+class RedisPubSubChannelLayer:
+    """
+    Channel Layer that uses Redis's pub/sub functionality.
+    """
+
+    def __init__(self, hosts, prefix="asgi", **kwargs):
+        assert (
+            isinstance(hosts, list) and len(hosts) > 0
+        ), "`hosts` must be a list with at least one Redis server"
+
+        self.prefix = prefix
+
+        # Each consumer gets its own *specific* channel, created with the `new_channel()` method.
+        # This dict maps `channel_name` to a queue of messages for that channel.
+        self.channels = {}
+
+        # A channel can subscribe to zero or more groups.
+        # This dict maps `group_name` to set of channel names who are subscribed to that group.
+        self.groups = {}
+
+        # For each host, we create a `RedisSingleShardConnection` to manage the connection to that host.
+        self._shards = [RedisSingleShardConnection(host, self) for host in hosts]
+
+    def _get_shard(self, channel_or_group_name):
+        """
+        Return the shard that is used exclusively for this channel or group.
+        """
+        if len(self._shards) == 1:
+            # Avoid the overhead of hashing and modulo when it is unnecessary.
+            return self._shards[0]
+        shard_index = abs(hash(channel_or_group_name)) % len(self._shards)
+        return self._shards[shard_index]
+
+    def _get_group_channel_name(self, group):
+        """
+        Return the channel name used by a group.
+        Includes '__group__' in the returned
+        string so that these names are distinguished
+        from those returned by `new_channel()`.
+        Technically collisions are possible, but it
+        takes what I believe is intentional abuse in
+        order to have colliding names.
+        """
+        return f"{self.prefix}__group__{group}"
+
+    extensions = ["groups"]
```

Comment:
OT: what is this for? When I search for its usage I cant find any. I can see it's defined in all other `ChannelLayer` classes too, but I cant figure out its purpose. Asgiref defined something like `extensions` in its docs but it seems just like a name conincidence (?)

Reply:
Honestly I only included it because I saw it in other `ChannelLayer`s (like you also noticed). E.g. [here](https://github.com/django/channels/blob/588f7d3c7d0734727ffdd358d5b48145b9979258/channels/layers.py#L217). Maybe someone, somewhere uses it...

Reply:
I saw it there too when I was searching for its purpose, but it doesnt seems to be used. I made it an empty list and everything was still working the same 🤔 

Maybe I'll make a MR to Channels to remove it from there to see what will be the reaction.